### PR TITLE
[java] NPathComplexity can't handle switch expressions

### LINKED ISSUE #2625: [java] NPathComplexity can't handle switch expressions
**Affects PMD Version:** 
6.25

**Rule**
[NPathComplexity](https://pmd.github.io/latest/pmd_rules_java_design.html#npathcomplexity)

**Description:**
When replacing java switch statements with the new switch expressions, NPathComplexity calculation goes through the roof. The following (obfuscated) code sample gives us an NPath complexity of 2147483645. Before the change to switch expressions, it was well below our threshold of 200.

**Code Sample demonstrating the issue:**

```java
public class NPathTest {
    Product buildProduct(String title, Language language) {
        if (title.length() > 100 || title.length() == 0) {
            throw new IllegalArgumentException("Invalid title length (1-100 chars): " + title);
        }
        val product = new Product();
        val description = new MultiLangString();
        switch (language) {
            case FR -> description.setFr(title);
            case IT -> description.setIt(title);
            case EN -> description.setEn(title);
            default -> description.setDe(title);
        }
        product.setDescription(description);
        product.setMagicNumber(3);
        return product;
    }
}
```

**Running PMD through:** *[CLI]*
JDK: OpenJDK 14

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

Add support for multiple case labels, switch expressions and switch blocks in NPATH and CYCLO metrics.

## Related issues

- Fixes #2625

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by adangel** [comment]
Do we need to mention this change on https://pmd.github.io/latest/pmd_java_metrics_index.html#cyclomatic-complexity-cyclo and https://pmd.github.io/latest/pmd_java_metrics_index.html#npath-complexity-npath ?

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/6ada7dec2bcd454abe852937c46d8643/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml [hunk 1]
```diff
@@ -389,6 +389,71 @@ class Foo {
...
+            <message>'Foo#bar()' has value 1.</message>
+            <message>'Foo#foo()' has value 5.</message>
+        </expected-messages>
+        <code><![CDATA[
+class Foo {
+    void bar(){}
+    void foo() {
+        int x=0;
+        boolean a, b;
+        switch (x) {                            // +1
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml:430]: Just being picky, the +1 is not for the switch, but for the method block

```suggestion
        switch (x) {
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml [hunk 2]
```diff
@@ -389,6 +389,71 @@ class Foo {
...
+        <description>Switch labeled blocks should count</description>
+        <expected-problems>2</expected-problems>
+        <expected-messages>
+            <message>'Foo#bar()' has value 1.</message>
+            <message>'Foo#foo()' has value 5.</message>
+        </expected-messages>
+        <code><![CDATA[
+class Foo {
+    void bar(){}
+    void foo() {
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml:427]:
```suggestion
    void foo() { // +1
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml [hunk 3]
```diff
@@ -389,6 +389,71 @@ class Foo {
...
+        int x=0;
+        switch (x) {                // +1
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml:451]: Same

```suggestion
        switch (x) {
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml [hunk 4]
```diff
@@ -389,6 +389,71 @@ class Foo {
...
+        <description>Switch labeled expressions should count</description>
+        <expected-problems>2</expected-problems>
+        <expected-messages>
+            <message>'Foo#bar()' has value 1.</message>
+            <message>'Foo#foo()' has value 5.</message>
+        </expected-messages>
+        <code><![CDATA[
+class Foo {
+    void bar(){}
+    void foo() {
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml:449]:
```suggestion
    void foo() { // +1
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml (hunk 1 above)
Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/CycloTest.xml:430]: I think, being picky here, is extremely useful :) This hopefully avoids later questions.... It's important, that the test case is correct (including the comments). So, the "+1" should be added to line 427 then?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/internal/CycloVisitor.java [hunk 5]
```diff
@@ -65,13 +67,24 @@ public Object visit(ASTSwitchStatement node, Object data) {
...
                     && node.getChild(label.getIndexInParent() + 1) instanceof ASTBlockStatement) {
                 // an empty label is only counted if we count boolean paths
                 ((MutableInt) data).increment();
             }
         }
+
+        for (ASTSwitchLabeledRule rule : node.findChildrenOfType(ASTSwitchLabeledRule.class)) {
+            ASTSwitchLabel label = rule.getFirstChildOfType(ASTSwitchLabel.class);
+            if (label.isDefault()) {
+                continue;
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/internal/CycloVisitor.java:81]: I'll try to add this case on Thursday....

**Comment by adangel** [comment]
@oowekyala  Thanks for the review! I think, that's ready now. Let me know, if you want to merge, or whether I should do it.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/internal/CycloVisitor.java [hunk 6]
```diff
@@ -65,13 +67,24 @@ public Object visit(ASTSwitchStatement node, Object data) {
...
             if (considerBooleanPaths) {
-                ((MutableInt) data).increment();
+                ((MutableInt) data).add(label.findChildrenOfType(ASTExpression.class).size());
             } else if (node.getNumChildren() > 1 + label.getIndexInParent()
                     && node.getChild(label.getIndexInParent() + 1) instanceof ASTBlockStatement) {
                 // an empty label is only counted if we count boolean paths
                 ((MutableInt) data).increment();
             }
         }
+
```
[the remark this answers is not in the document]
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/internal/CycloVisitor.java:77]: I think the other test case was not sufficient. Eg the following fails, the complexity of the switch expr is not counted:
```java
class Foo {
    int bar() { return 1; }
    void foo(int i) {                    // +1
        int x = switch (i) {
            case 1, 2, 3 -> bar();  // +3
            case 4       -> bar();  // +1
            default -> bar();
        };
    }
}
```

I'll push a fix

**Pull request merged**